api/src/main/java/me/lokka30/treasury/api/economy/currency/Currency.java

```diff
@@ -12,69 +12,216 @@
+     * always returning 0.
+     *
+     * @param currencyChar          a way to identify the currency via a character e.g. '$' for a dollar. can be null.
+     * @param roundedDigits         to which digit shall we round balances. can specify -1 for no rounding.
+     * @param conversionCoefficient coefficient at which this currency is going to be converted to other currencies.
+     * @param balanceFormatter      a function, giving a human-readable format of a balance.
+     * @param names                 acceptable names of the currency. primary name shall be the first specified.
+     * @return new currency instance
+     */
+    @NotNull
+    public static Currency of(
+            @Nullable Character currencyChar,
+            int roundedDigits,
+            double conversionCoefficient,
+            @NotNull BiFunction<Double, Locale, String> balanceFormatter,
+            @NotNull String @NotNull... names
+    ) {
+        return new Currency(
+                null,
+                currencyChar,
+                roundedDigits,
+                conversionCoefficient,
+                null,
+                balanceFormatter,
+                names
+        );
+    }
+
+    /**
+     * Creates a new currency.
+     *
+     * @param currencyId            the currency id.
+     * @param currencyChar          a way to identify the currency via a character e.g. '$' for a dollar. can be null.
+     * @param roundedDigits         to which digit shall we round balances. can specify -1 for no rounding.
+     * @param conversionCoefficient coefficient at which this currency is going to be converted to other currencies.
+     * @param startingBalance       a function, giving a starting balance for the inputted player uuid, which can be null.
+     * @param balanceFormatter      a function, giving a human-readable format of a balance.
+     * @param names                 acceptable names of the currency. primary name shall be the first specified.
+     * @return new currency instance
+     */
+    @NotNull
+    public static Currency of(
+            @NotNull UUID currencyId,
+            @Nullable Character currencyChar,
+            int roundedDigits,
+            double conversionCoefficient,
+            @Nullable Function<UUID, Double> startingBalance,
+            @NotNull BiFunction<Double, Locale, String> balanceFormatter,
+            @NotNull String @NotNull... names
+    ) {
+        Objects.requireNonNull(currencyId, "currencyId");
+        return new Currency(
+                currencyId,
+                currencyChar,
+                roundedDigits,
+                conversionCoefficient,
+                startingBalance,
+                balanceFormatter,
+                names
+        );
+    }
+
+    private final UUID currencyId;
+    private final Character currencyChar;
+    private final String[] names;
+    private final int roundedDigits;
+    private final double conversionCoefficient;
```

Comment:
Can we make this protected and not final? Or a protected setter?

Reply:
I don't understand why. I don't even understand why you want to extend Currency. Sure thing, you want to do cryptocurrencies and you for some reason need to change the conversion coefficient. Just do it via a config of some kind and make your server restart every 12 hours to apply the changes or every day at 12 midnight. 

The given value is and shall not be controlled by the conversion coefficient, unless you _really_ need to convert from currency x to currency y. You can fetch every day from the internet the value of the bitcoin and do calculations based on that in order to determine what amount shall you give them. 

I'm starting to think to return the `final` modifier on the class too...

Reply:
Why make end users restart to reload a configuration? That will only encourage use of `/reload`. There's also no real harm in allowing extension of currencies, though I agree it doesn't seem like a particularly great thing to do. Allowing on-the-fly currency modification would definitely make it more possible to do stock exchange style things without having to unregister (which, as much as there's no API for it, is certainly possible with reflection) and re-register the currency.

Reply:
@Jikoo @MrNemo64 happy now!?